File: backend/app/schemas/course_default_rubric.py
"""Schemas for course-level default rubric (JSON storage + API)."""

from __future__ import annotations

from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field, model_validator


class CourseDefaultRubricCriterion(BaseModel):
    name: str = Field(min_length=1)
    description: str = ""
    maxPoints: int = Field(default=5, ge=0, le=1000)
    weight: float = Field(default=10.0, ge=0, le=100)
    gradingMethod: Literal["auto", "manual", "hybrid"] = "manual"
    defaultComments: Optional[Dict[str, str]] = None


class CourseDefaultRubricSection(BaseModel):
    name: str = Field(min_length=1)
    description: str = ""
    weight: float = Field(ge=0, le=1000)
    criteria: List[CourseDefaultRubricCriterion] = Field(default_factory=list)


class CourseDefaultRubricPut(BaseModel):
    rubricMode: Literal["weighted", "unweighted"] = "weighted"
    weightPolicy: Literal["percent", "points"] = "percent"
    pointBudget: float = Field(default=100.0, gt=0, le=10_000)
    sections: List[CourseDefaultRubricSection]
    autoNormalize: bool = True

    @model_validator(mode="after")
    def require_sections(self):
        if not self.sections:
            raise ValueError("sections must contain at least one rubric section")
        for sec in self.sections:
            if not sec.criteria:
                raise ValueError(f'Section "{sec.name}" must include at least one criterion')
        return self


class CourseDefaultRubricOut(BaseModel):
    rubricMode: str = "weighted"
    weightPolicy: str = "percent"
    pointBudget: float = 100.0
    sections: List[CourseDefaultRubricSection]
    isBuiltin: bool = False
    updatedAt: Optional[datetime] = None
    updatedByUserId: Optional[int] = None
    updatedByName: Optional[str] = None


class AssignmentRubricReplaceBody(BaseModel):
    """Replace all rubric sections/criteria on an assignment."""

    rubricMode: Optional[Literal["weighted", "unweighted"]] = None
    rubric: List[CourseDefaultRubricSection] = Field(min_length=1)

    @model_validator(mode="after")
    def criteria_nonempty(self):
        for sec in self.rubric:
            if not sec.criteria:
                raise ValueError(f'Section "{sec.name}" must include at least one criterion')
        return self

File: frontend/src/lib/courseDefaultRubric.ts
export type RubricWeightPolicy = 'percent' | 'points';

export interface RubricFormCriterion {
    name: string;
    description: string;
    maxPoints: number;
    weight: number;
    gradingMethod: 'auto' | 'manual' | 'hybrid';
    defaultComments?: Record<string, string>;
}

export interface RubricFormSection {
    name: string;
    description: string;
    weight: number;
    criteria: RubricFormCriterion[];
}

export interface CourseDefaultRubricCriterionApi {
    name: string;
    description: string;
    maxPoints: number;
    weight: number;
    gradingMethod: 'auto' | 'manual' | 'hybrid';
    defaultComments?: Record<string, string> | null;
}

export interface CourseDefaultRubricSectionApi {
    name: string;
    description: string;
    weight: number;
    criteria: CourseDefaultRubricCriterionApi[];
}

export interface CourseDefaultRubricApi {
    rubricMode: 'weighted' | 'unweighted';
    weightPolicy: RubricWeightPolicy;
    pointBudget: number;
    sections: CourseDefaultRubricSectionApi[];
    isBuiltin: boolean;
    updatedAt?: string | null;
    updatedByUserId?: number | null;
    updatedByName?: string | null;
}

export interface CourseDefaultRubricPutApi {
    rubricMode: 'weighted' | 'unweighted';
    weightPolicy: RubricWeightPolicy;
    pointBudget: number;
    sections: CourseDefaultRubricSectionApi[];
    autoNormalize: boolean;
}

/** Grade scale used across the rubric */
export const GRADE_SCALE = [0, 1, 2, 3, 4, 5] as const;
export const MAX_GRADE = 5;

/** Calculate points for a criterion: (grade / maxGrade) × weight */
export function criterionPoints(grade: number, weight: number): number {
    return Math.round(((grade / MAX_GRADE) * weight) * 100) / 100;
}

/** Convert course default API response → Create Assignment form slice. */
export function courseDefaultApiToFormPartial(api: CourseDefaultRubricApi): {
    rubricMode: 'weighted' | 'unweighted';
    rubricWeightKind: RubricWeightPolicy;
    rubric: RubricFormSection[];
} {
    const rubric: RubricFormSection[] = api.sections.map((s) => ({
        name: s.name,
        description: s.description ?? '',
        weight: s.weight,
        criteria: (s.criteria ?? []).map((c) => ({
            name: c.name,
            description: c.description ?? '',
            maxPoints: c.maxPoints ?? 5,
            weight: c.weight,
            gradingMethod: c.gradingMethod,
            defaultComments: c.defaultComments ?? undefined,
        })),
    }));

    return {
        rubricMode: api.rubricMode,
        rubricWeightKind: api.weightPolicy,
        rubric,
    };
}

export function formRubricToCoursePutApi(data: {
    rubric: RubricFormSection[];
    rubricMode: 'weighted' | 'unweighted';
    rubricWeightKind?: RubricWeightPolicy;
    maxPoints: number;
}): CourseDefaultRubricPutApi {
    const weightPolicy = data.rubricWeightKind ?? 'percent';
    const pointBudget = data.maxPoints > 0 ? data.maxPoints : 100;
    return {
        rubricMode: data.rubricMode,
        weightPolicy,
        pointBudget,
        sections: (data.rubric ?? []).map((s) => ({
            name: s.name,
            description: s.description ?? '',
            weight: s.weight ?? 0,
            criteria: (s.criteria ?? []).map((c) => ({
                name: c.name,
                description: c.description ?? '',
                maxPoints: c.maxPoints ?? 5,
                weight: typeof c.weight === 'number' ? c.weight : 0,
                gradingMethod: c.gradingMethod,
                defaultComments: c.defaultComments ?? null,
            })),
        })),
        autoNormalize: true,
    };
}
